src/uri.cr

```diff
@@ -144,6 +144,16 @@ class URI
     end
   end
 
+  # Returns true if URI has a scheme (e.g. http:// or https://) specified.
```

Comment:
A URI is only considered absolute if it specifies a scheme **and** hierarchical part (authority and/or path). For example, `http://#foo` is a relative URI. The implementation needs to take that into account as well.

Reply:
Just curious is `http://#foo` valid URL?

Reply:
Also `URI.parse("http://#foo").host # => ""` empty string.

Is it will OK?

```crystal
def absolute?
  @scheme && !@host.try(&.empty?) ? true : false
end
```

Reply:
> Just curious is `http://#foo` valid URL?

The question should be, is it a valid URI. And yes, it is. Obviously, it doesn't make much sense, but serves to illustrate the issue.

`@scheme && !@host.try(&.empty?)` seems to work with `URI.parse`. But it looks like this depends on an implementation detail of the parser (`host` is `nil` if path is non-empty, but `""` if path is empty).

Example:
```cr
URI.parse("file://").absolute? # => false
URI.new("file").absolute? # => true
```

You should add `&& !@path.try(&.empty?)` to the condition.

But remove `? true : false`. It doesn't do anything.

Reply:
It does:

```crystal
true && nil
# => nil
```

"https://example.com" is valid absolute URL, and `path` is empty.

There are too many cases to check if URL is absolute. I wouldn't like to overload the logic.

In most cases, the first variant is enought.

Reply:
Let's copy what everyone else is doing:
- [Ruby](https://github.com/ruby/ruby/blob/trunk/lib/uri/generic.rb#L962-L968)
- [Go](https://golang.org/src/net/url/url.go?s=26128:26154#L930)

Just checking whether `scheme` is set or not seems to be enough (like in the original PR).

Reply:
@asterite In most cases it is enough because URIs with scheme typically also include a authority or path. But that doesn't make it *correct*.

Reply:
@straight-shoota Let me quote the RFC in question: 

"Relative URI references are distinguished from absolute URI in that
   they do not begin with a scheme name."

Going by that checking the scheme is both correct and sufficient.

But there are some additional details, see section 5.2 of https://www.ietf.org/rfc/rfc2396.txt

Reply:
Seems to be okay than. Sorry for the confusion. There are different definitions of the same terms... but following RFC 2396 is probably the best 👍 

@mamantoha Would you change it back, please?